### PR TITLE
Add project endpoints

### LINKED ISSUE #218: [BE Feat] Create CreateProject project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [x] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #219: [BE Feat] Create GetProject project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #220: [BE Feat] Create GetProjects project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #221: [BE Feat] Create PatchProjectAnswer project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #222: [BE Feat] Create UploadProjectDocument project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #224: [BE Feat] Create DeleteProjectDocument project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #225: [BE Feat] Create SubmitProjectForReview project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [ ] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### LINKED ISSUE #226: [BE Feat] Create ListCompanyProjects project endpoint
## Checklist
Before submitting this ticket, make sure this ticket:
- [x] Is **labelled properly** - if this is a refactor, add the label!
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

And lastly:
- [x] Once someone picks up this ticket, **make sure to assign it**.

## Description
Refer to the Roadmap document on Notion.

### PR DESCRIPTION
**Pull request by AmirAgassi**
## Description
Added all project related endpoints including create, get, list, submit for review, upload/delete documents, and patch answers. Added validation for required fields and file uploads, with proper error handling and response formatting.

## Linked Issues
- Fixes #218 
- Fixes #219 
- Fixes #220 
- Fixes #221 
- Fixes #222 
- Fixes #224 
- Fixes #225 
- Fixes #226 

## Testing
- `backend\internal\tests\server_test.go`

## Checklist
Before opening this PR, make sure the PR:
- [x] Has an **assignee or group of assignees**.
- [x] Has a **reviewer or a group of reviewers**.
- [x] Is **labelled properly**.
- [x] Has the **SPUR project assigned to it**.
- [x] Has an **assigned milestone**.

Additionally, make sure that:
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: backend/.sqlc/migrations/20241226195458_add_default_questions.sql
Comment (juancwu) [inline backend/.sqlc/migrations/20241226195458_add_default_questions.sql:1]: Is this meant to be a seeding for the question? And if yes, is this a legit question from Notion? Asking cuz I won't include these in #189

File: backend/.sqlc/queries/projects.sql [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    id,
+    company_id,
+    title,
+    description,
+    status,
+    created_at,
+    updated_at
+) VALUES (
+    $1, $2, $3, $4, $5, $6, $7
+) RETURNING *;
```
Comment (juancwu) [inline backend/.sqlc/queries/projects.sql]: The `id` column should have a default generation for the uuid so it's not necessary to generate it ourselves when inserting a new row.

File: backend/.sqlc/queries/projects.sql [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    answer = $1,
+    updated_at = extract(epoch from now())
+WHERE 
+    project_answers.id = $2 
+    AND project_id = $3 
+    AND project_id IN (
+        SELECT projects.id 
+        FROM projects 
+        WHERE projects.company_id = $4
+    )
```
Comment (juancwu) [inline backend/.sqlc/queries/projects.sql]: I don't get this sub-query... I think this is extra work since a project answer `id` is unique so it should always just belong to the given project and company. I think its better to check the relationship in Go instead if it is needed.

File: backend/internal/v1/v1_projects/projects.go [hunk 3]
```diff
@@ -1 +1,724 @@
...
+		Status:      db.ProjectStatusDraft,
+		CreatedAt:   now,
+		UpdatedAt:   now,
+	})
+	if err != nil {
+		return v1_common.Fail(c, 500, "Failed to create project", err)
+	}
+
+	// Create empty answers within same transaction
+	_, err = qtx.CreateProjectAnswers(ctx, project.ID)
```
Comment (juancwu) [inline backend/internal/v1/v1_projects/projects.go]: I don't think we need to do this at all at this step ngl. The frontend can be sending information of the question id and project id that incrementally add answers instead. I think that pattern removes the need to use a transaction too.

File: backend/internal/tests/server_test.go [hunk 4]
```diff
@@ -383,19 +413,40 @@ func TestServer(t *testing.T) {
...
+			rec := httptest.NewRecorder()
+			s.Echo.ServeHTTP(rec, req)
+			assert.Equal(t, http.StatusCreated, rec.Code)
+
+			// Get user from database
+			var user db.User
+			err = s.DBPool.QueryRow(ctx, "SELECT id FROM users WHERE email = $1", email).Scan(&user.ID)
+			assert.NoError(t, err)
+
+			// Generate test email token
```
Comment (juancwu) [inline backend/internal/tests/server_test.go:441]: Why was this changed? Isn't the helper methods doing the same thing?

File: backend/internal/tests/server_test.go
Comment (juancwu) [inline backend/internal/tests/server_test.go:1]: Was the helper function to create/remove test user causing issues? Just being extra careful since changes to tests that are not really part of the projects routes doesn't seem right to me

File: backend/internal/v1/v1_projects/types.go [hunk 5]
```diff
@@ -1 +1,71 @@
...
+
+type PatchAnswerRequest struct {
+	Content  string `json:"content" validate:"required"`
+	AnswerID string `json:"answer_id" validate:"required,uuid"`
+}
+
+type UploadDocumentRequest struct {
+	Name    string `json:"name" validate:"required"`
+	Section string `json:"section" validate:"required"`
+}
```
Comment (juancwu) [inline backend/internal/v1/v1_projects/types.go:42]: I think the project id should also be included, otherwise, we won't be able to know which project the document belongs to, right?

File: backend/internal/middleware/jwt.go [hunk 6]
```diff
@@ -37,47 +38,50 @@ func AuthWithConfig(config AuthConfig, dbPool *pgxpool.Pool) echo.MiddlewareFunc
...
+				}
+			}
+
 			// store claims and user in context for handlers
 			c.Set("claims", &AuthClaims{
 				JWTClaims: claims,
 				Salt:      user.TokenSalt,
 			})
 			c.Set("user", &user)
+			c.Set("user_id", uuid.MustParse(claims.UserID))
```
Comment (juancwu) [inline backend/internal/middleware/jwt.go]: This is sort of redundant because you can get the user id from the user struct. This adds extra memory usage

File: backend/internal/v1/v1_projects/projects.go [hunk 7]
```diff
@@ -1 +1,724 @@
...
+ * handlePatchProjectAnswer updates an answer for a project question.
+ * 
+ * Validation:
+ * - Validates answer content against question rules
+ * - Returns validation errors if content invalid
+ * 
+ * Security:
+ * - Verifies project belongs to user's company
+ */
+func (h *Handler) handlePatchProjectAnswer(c echo.Context) error {
```
Comment (juancwu) [inline backend/internal/v1/v1_projects/projects.go:188]: We should check the project status is 'draft' before updating the answer.

File: backend/internal/v1/v1_projects/projects.go [hunk 8]
```diff
@@ -1 +1,724 @@
...
+	// Get company owned by user
+	company, err := h.server.GetQueries().GetCompanyByUserID(c.Request().Context(), userID.String())
+	if err != nil {
+		return v1_common.Fail(c, http.StatusNotFound, "Company not found", err)
+	}
+
+	projectID := c.Param("id")
+	if projectID == "" {
+		return v1_common.Fail(c, http.StatusBadRequest, "Project ID is required", nil)
+	}
```
Comment (juancwu) [inline backend/internal/v1/v1_projects/projects.go:630]: I find verifying static things before doing any query is generally better. Removes the need to hit the database if the request is not good

File: backend/internal/v1/v1_projects/projects.go
Comment (juancwu) [inline backend/internal/v1/v1_projects/projects.go:1]: Authorization not checked. I think only the owners and the admins should have access to view/edit projects

File: backend/internal/v1/v1_projects/types.go (hunk 5 above)
Comment (juancwu) [reply backend/internal/v1/v1_projects/types.go:42]: nvm this, the project id is gotten from the url

File: backend/.sqlc/migrations/20241226195458_add_default_questions.sql
Comment (AmirAgassi) [reply backend/.sqlc/migrations/20241226195458_add_default_questions.sql:1]: Yup, I was thinking I'd make the seed migration for the questions in this PR and you'd be able to edit the actual full list of questions (once Nausher finishes 'em) in it later.

These aren't actual questions from the Notion though, just a few random ones to be able to test answering questions and stuff.

**Comment by juancwu** [comment]
Overall, the only important change is to add authorization checks for the actions

**Review by juancwu — APPROVED** [review]
lgtm
